ibis/pandas/aggcontext.py

```diff
@@ -252,19 +252,41 @@ def agg(self, grouped_data, function, *args, **kwargs):
         pass
 
 
-def make_applied_function(function, args=None, kwargs=None):
+def wrap_for_apply(function, args, kwargs):
+    """Wrap a function so that it may be used with Pandas `apply`"""
     assert callable(function), 'function {} is not callable'.format(function)
 
     @functools.wraps(function)
-    def apply(
-        data,
-        function=function,
-        args=args if args is not None else (),
-        kwargs=kwargs if kwargs is not None else {},
-    ):
+    def wrapped_func(data, function=function, args=args, kwargs=kwargs):
+        return function(data, *args, **kwargs)
+
+    return wrapped_func
+
+
+def wrap_for_agg(function, args, kwargs):
+    """Wrap a function so that it may be used with Pandas `agg`.
+
+    In some cases, Pandas `agg` will try to behave like Pandas `apply`.
+    `function` will be wrapped in such a way that Pandas will be forced
+    not to do this (in other words, forced to treat `function` as an
+    aggregation function, not a mapping function).
+    """
+    assert callable(function), 'function {} is not callable'.format(function)
+
+    @functools.wraps(function)
+    def wrapped_func(data, function=function, args=args, kwargs=kwargs):
+        # `data` will be a scalar here if Pandas `agg` is trying to behave like
+        # like Pandas `apply`.
+        if not isinstance(data, pd.Series):
```

Comment:
Why do we enforce pd.Series here? Can this take np.ndarray?

Reply:
This should be a `pd.Series` always

This function (`wrapped_func`) will be called by `pandas.Series.aggregate` [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/series.py#L4038), which will pass `self` for the first argument. Since `self` is a `pd.Series`, `data` would be a `pd.Series`.

Reply:
Ok. This seems fine. Asking this because I know what aggcontext.Window can take numpy array as data.

Reply:
Oh that reminds me of one detail—`aggcontext.Summarize` takes either a `Series` or a `SeriesGroupBy` for the `grouped_data` argument
https://github.com/ibis-project/ibis/blob/3ba1751f09dd7d4b8af74c3c1ef8b63bd7581760/ibis/pandas/aggcontext.py#L295-L304

https://github.com/ibis-project/ibis/blob/3ba1751f09dd7d4b8af74c3c1ef8b63bd7581760/ibis/pandas/aggcontext.py#L304

In the line above, `grouped_data` can either be a `Series` **or a `SeriesGroupBy`**.

If `grouped_data` is a `SeriesGroupBy`, Pandas will "unpack" each group and pass a `Series` to `wrapped_func`.

So in either case, the `data` argument in `wrapped_func` will be a `Series`.